File: frontend/lib/execution-logs.ts
import { appConfig } from "@/lib/config";

export type ProjectInfo = {
  analysis_id: string;
  project_name: string;
  uploaded_at: string;
};

export type ExecutionLogUploadResponse = {
  id: string;
  project_id: string;
  project_name: string;
  filename: string;
  original_filename: string;
  size_bytes: number;
  status: string;
  entries_count: number;
  execution_records_count: number;
  duplicate_count: number;
  log_date_from: string | null;
  log_date_to: string | null;
  validation_messages: Array<{ field: string; message: string; severity: string }>;
  upload_date: string;
  total_log_lines: number;
  csv_rows_read: number;
  execution_starts_found: number;
  execution_ends_found: number;
  parse_errors: number;
  validation_warnings: number;
};

export type ExecutionRecordData = {
  project_id: string;
  project_name: string;
  workspace_name: string;
  environment_name: string;
  plan_name: string;
  artifact_name: string;
  task_execution_id: string;
  plan_execution_id: string;
  remote_engine_name: string;
  execution_start_time: string | null;
  execution_end_time: string | null;
  execution_status: string;
  execution_duration_seconds: number | null;
  error_message: string;
  original_log_file_name: string;
  upload_date: string | null;
  source_file: string;
};

export type ProjectUploadSummary = {
  project_id: string;
  project_name: string;
  upload_id: string;
  original_filename: string;
  upload_date: string | null;
  status: string;
  entries_count: number;
  execution_records_count: number;
  log_date_from: string | null;
  log_date_to: string | null;
  validation_messages: Array<{ field: string; message: string; severity: string }>;
  total_log_lines: number;
  csv_rows_read: number;
  execution_starts_found: number;
  execution_ends_found: number;
  parse_errors: number;
  validation_warnings: number;
};

export type ExecutionLogHistoryItem = {
  id: string;
  project_name: string;
  project_id: string;
  filename: string;
  original_filename: string;
  upload_date: string;
  processing_status: string;
  size_bytes: number;
};

export async function fetchProjects(): Promise<ProjectInfo[]> {
  const response = await fetch(
    `${appConfig.apiBaseUrl.replace(/\/$/, "")}/execution-logs/projects`,
    { cache: "no-store" },
  );
  if (!response.ok) {
    throw new Error("Failed to fetch projects.");
  }
  return response.json() as Promise<ProjectInfo[]>;
}

export function uploadExecutionLog(
  projectId: string,
  projectName: string,
  file: File,
  onProgress: (progress: number) => void,
): Promise<ExecutionLogUploadResponse> {
  return new Promise((resolve, reject) => {
    const request = new XMLHttpRequest();
    const formData = new FormData();
    formData.append("project_id", projectId);
    formData.append("project_name", projectName);
    formData.append("file", file, file.name);

    request.upload.onprogress = (event) => {
      if (!event.lengthComputable) return;
      onProgress(Math.round((event.loaded / event.total) * 100));
    };

    request.onload = () => {
      const responseText = request.responseText || "{}";
      if (request.status === 201) {
        onProgress(100);
        resolve(JSON.parse(responseText) as ExecutionLogUploadResponse);
        return;
      }
      try {
        const parsed = JSON.parse(responseText) as { detail?: string };
        reject(new Error(parsed.detail ?? `Upload failed with status ${request.status}.`));
      } catch {
        reject(new Error(`Upload failed with status ${request.status}.`));
      }
    };

    request.onerror = () => {
      reject(new Error("Unable to reach the upload API."));
    };

    request.open(
      "POST",
      `${appConfig.apiBaseUrl.replace(/\/$/, "")}/execution-logs/upload`,
    );
    request.send(formData);
  });
}

export async function fetchUploadHistory(): Promise<ExecutionLogHistoryItem[]> {
  const response = await fetch(
    `${appConfig.apiBaseUrl.replace(/\/$/, "")}/execution-logs/history`,
    { cache: "no-store" },
  );
  if (!response.ok) {
    throw new Error("Failed to fetch upload history.");
  }
  return response.json() as Promise<ExecutionLogHistoryItem[]>;
}

export async function fetchExecutionRecords(
  projectId: string,
): Promise<ExecutionRecordData[]> {
  try {
    const response = await fetch(
      `${appConfig.apiBaseUrl.replace(/\/$/, "")}/execution-logs/projects/${projectId}/records`,
      { cache: "no-store" },
    );
    if (!response.ok) return [];
    return (await response.json()) as ExecutionRecordData[];
  } catch {
    return [];
  }
}

export async function fetchProjectUploadSummary(
  projectId: string,
): Promise<ProjectUploadSummary | null> {
  try {
    const response = await fetch(
      `${appConfig.apiBaseUrl.replace(/\/$/, "")}/execution-logs/projects/${projectId}/summary`,
      { cache: "no-store" },
    );
    if (!response.ok) return null;
    return (await response.json()) as ProjectUploadSummary;
  } catch {
    return null;
  }
}

File: backend/schemas/execution_logs.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel


class ProjectInfo(BaseModel):
    analysis_id: str
    project_name: str
    uploaded_at: datetime


class ExecutionLogUploadResponse(BaseModel):
    id: str
    project_id: str
    project_name: str = ""
    filename: str
    original_filename: str
    size_bytes: int
    status: str
    entries_count: int = 0
    execution_records_count: int = 0
    duplicate_count: int = 0
    log_date_from: datetime | None = None
    log_date_to: datetime | None = None
    validation_messages: list[dict[str, Any]] = []
    upload_date: datetime | None = None
    total_log_lines: int = 0
    csv_rows_read: int = 0
    execution_starts_found: int = 0
    execution_ends_found: int = 0
    parse_errors: int = 0
    validation_warnings: int = 0


class ExecutionLogHistoryItem(BaseModel):
    id: str
    project_name: str
    project_id: str
    filename: str
    original_filename: str
    upload_date: datetime
    processing_status: str
    size_bytes: int
    entries_count: int = 0


class ExecutionLogEntryResponse(BaseModel):
    job_name: str | None = None
    execution_id: str | None = None
    start_time: datetime | None = None
    end_time: datetime | None = None
    status: str | None = None
    duration_seconds: float | None = None
    error_message: str | None = None
    restart_time: datetime | None = None
    environment: str | None = None


class ExecutionLogUploadDetail(BaseModel):
    id: str
    project_id: str
    project_name: str
    filename: str
    original_filename: str
    size_bytes: int
    upload_date: datetime
    status: str
    error: str | None = None
    entries: list[ExecutionLogEntryResponse] = []


class ExecutionRecordResponse(BaseModel):
    project_id: str = ""
    project_name: str = ""
    workspace_name: str = ""
    environment_name: str = ""
    plan_name: str = ""
    artifact_name: str = ""
    task_execution_id: str = ""
    plan_execution_id: str = ""
    remote_engine_name: str = ""
    execution_start_time: datetime | None = None
    execution_end_time: datetime | None = None
    execution_status: str = ""
    execution_duration_seconds: float | None = None
    error_message: str = ""
    restart_time: datetime | None = None
    original_log_file_name: str = ""
    upload_date: datetime | None = None
    execution_attempt: int = 1
    source_file: str = ""


class ProjectUploadSummary(BaseModel):
    project_id: str
    project_name: str = ""
    upload_id: str
    original_filename: str
    upload_date: datetime | None = None
    status: str = ""
    entries_count: int = 0
    execution_records_count: int = 0
    log_date_from: datetime | None = None
    log_date_to: datetime | None = None
    validation_messages: list[dict[str, Any]] = []
    total_log_lines: int = 0
    csv_rows_read: int = 0
    execution_starts_found: int = 0
    execution_ends_found: int = 0
    parse_errors: int = 0
    validation_warnings: int = 0
